### PR TITLE
Implement provisional identities

### PR DESCRIPTION
**Pull request by Marwyn42**

### CODE REVIEW

File: identity/generate.go [hunk 1]
```diff
@@ -1,145 +1,32 @@
...
-//Exported only to facilitate code testing (this shouldn't be needed in any app using this lib)
-type SecretPermanentIdentity struct {
-	PublicPermanentIdentity
-	DelegationSignature          []byte `json:"delegation_signature"`
-	EphemeralPublicSignatureKey  []byte `json:"ephemeral_public_signature_key"`
-	EphemeralPrivateSignatureKey []byte `json:"ephemeral_private_signature_key"`
-	UserSecret                   []byte `json:"user_secret"`
-}
-
 //Create a user token for given user.
```
Comment (banetl) [inline identity/generate.go]: Also change the comment here

File: identity/b64json.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package identity
```
Comment (banetl) [inline identity/b64json.go]: Not sure that this file belongs in identity, might be better to put it in separate package

File: identity/identity.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	EphemeralPublicSignatureKey  []byte `json:"ephemeral_public_signature_key"`
+	EphemeralPrivateSignatureKey []byte `json:"ephemeral_private_signature_key"`
+	UserSecret                   []byte `json:"user_secret"`
+}
+
+func generateIdentity(config config, userIDString string) (*identity, error) {
+	userID := hashUserID(config.TrustchainID, userIDString)
+	userSecret := createUserSecret(userID)
+
+	eprivSignKey, epubSignKey, err := ed25519.GenerateKey(nil)
```
Comment (banetl) [inline identity/identity.go]: is it possible to import ed25519 as crypto and call `crypto.GenerateKey(nil)` ?

File: identity/user_secret.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package identity
+
+import (
+	"crypto/rand"
+
+	"golang.org/x/crypto/blake2b"
+)
+
+const userSecretSize = 32
```
Comment (banetl) [inline identity/user_secret.go]: shouldn't you get this value from a constant in crypto ?

File: identity/user_secret.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+func hashUserID(trustchainID []byte, userIDString string) []byte {
+	userIDBuffer := append([]byte(userIDString), trustchainID...)
+	hashedUserID := blake2b.Sum256(userIDBuffer)
+	return hashedUserID[:]
+}
+
+func createUserSecret(userID []byte) []byte {
+	randdata := make([]byte, userSecretSize-1)
```
Comment (banetl) [inline identity/user_secret.go]: this is a bad function name, add a comment when you use cryptic functions such as this one

File: identity/b64json.go (hunk 2 above)
Comment (Marwyn42) [reply identity/b64json.go]: Done

File: identity/identity.go (hunk 3 above)
Comment (Marwyn42) [reply identity/identity.go]: Nope. This is the name of the algorithm we generate a key for

File: curve25519/curve25519Generate_test.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	It("generates a non nil key pair", func() {
+		sk, pk, err := curve25519.GenerateKey()
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(len(sk)).Should(Equal(32))
+		Expect(len(pk)).Should(Equal(32))
+
+		emptyArray := [32]byte{}
+		Expect(sk).ShouldNot(Equal(emptyArray[:]))
+		Expect(pk).ShouldNot(Equal(emptyArray[:]))
+	})
```
Comment (tux3) [inline curve25519/curve25519Generate_test.go]: You could test that if you generate two keys, they're actually different, just as a sanity check!

Comment (Marwyn42) [reply curve25519/curve25519Generate_test.go]: Done

**Pull request merged**